### PR TITLE
[Feature] follow-up work to enable exporting to onnx for PointRend (#4977)

### PR DESCRIPTION
**Pull request by AronLin**
This PR is the follow-up work to enable exporting to ONNX for PointRend.
The origin PR is #4977  and open-mmlab/mmcv#953.

## Motivation

The origin PR completes the exporting work but we refactor the code logic for ONNX exporting.
So I completed the follow-up logic reconstruction work for PointRend.

## Test

Here is the result:

|    Model     |  Type   | box mAP | box AP50| box AP75| mask mAP | mask AP50 | mask AP75 |
:--------------: | :------: | :------------: | :----: | :-----: | :------: | :--------: | :--------: |
point_rend_r50_caffe_fpn_mstrain_1x_coco| PyTorch | 38.4 | 59.0 | 41.8 | 36.3 | 56.9 | 38.7 |
point_rend_r50_caffe_fpn_mstrain_1x_coco| ONNX | 38.4 | 59.0 | 41.8 | 35.2 | 56.5 | 37.1 |

It seems all right.

### CODE REVIEW

File: mmdet/models/roi_heads/mask_heads/mask_point_head.py [hunk 1]
```diff
@@ -285,16 +285,20 @@ def get_roi_rel_points_test(self, mask_pred, pred_label, cfg):
...
+        mask_size = int(mask_height * mask_width)
+        uncertainty_map = uncertainty_map.view(num_rois, mask_size)
+        num_points = min(mask_size, num_points)
         point_indices = uncertainty_map.topk(num_points, dim=1)[1]
-        point_coords = uncertainty_map.new_zeros(num_rois, num_points, 2)
-        point_coords[:, :, 0] = w_step / 2.0 + (point_indices %
-                                                mask_width).float() * w_step
-        point_coords[:, :, 1] = h_step / 2.0 + (point_indices //
-                                                mask_width).float() * h_step
+        xs = w_step / 2.0 + (point_indices.long() %
```
Comment (jshilong) [inline mmdet/models/roi_heads/mask_heads/mask_point_head.py:299]: Isn't  type  of  `topk` return value originally `long`

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 2]
```diff
@@ -207,3 +208,153 @@ def aug_test_mask(self, feats, img_metas, det_bboxes, det_labels):
                 scale_factor=1.0,
                 rescale=False)
         return segm_result
+
+    def _onnx_get_fine_grained_point_feats(self, x, rois, rel_roi_points):
```
Comment (jshilong) [inline mmdet/models/roi_heads/point_rend_roi_head.py:212]: add docstr

Comment (jshilong) [reply mmdet/models/roi_heads/point_rend_roi_head.py:212]: add comments for the args

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 3]
```diff
@@ -207,3 +208,153 @@ def aug_test_mask(self, feats, img_metas, det_bboxes, det_labels):
...
+        return refined_mask_pred
+
+    def mask_onnx_export(self,
+                         x,
+                         img_metas,
+                         det_bboxes,
+                         det_labels,
+                         rescale=False,
+                         **kwargs):
+        """Export mask branch to onnx which supports batch inference.
```
Comment (jshilong) [inline mmdet/models/roi_heads/point_rend_roi_head.py:296]: test the batch inference

File: mmdet/models/roi_heads/mask_heads/mask_point_head.py [hunk 4]
```diff
@@ -285,16 +285,20 @@ def get_roi_rel_points_test(self, mask_pred, pred_label, cfg):
...
+                             mask_width).float() * w_step
```
Comment (jshilong) [inline mmdet/models/roi_heads/mask_heads/mask_point_head.py:300]: If this modification is involved in inference, please check the checkpoints of `point_rend`

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 5]
```diff
@@ -207,3 +208,153 @@ def aug_test_mask(self, feats, img_metas, det_bboxes, det_labels):
...
+                                                       mask_height, mask_width)
+
+        return refined_mask_pred
+
+    def mask_onnx_export(self,
+                         x,
+                         img_metas,
+                         det_bboxes,
+                         det_labels,
+                         rescale=False,
```
Comment (jshilong) [inline mmdet/models/roi_heads/point_rend_roi_head.py:294]: Seems `rescale` would never be used when doing the `onnx_export`, you may delete all the `rescale` related codes. But it is best to confirm by  @RunningLeon

File: mmdet/models/roi_heads/mask_heads/mask_point_head.py [hunk 6]
```diff
@@ -285,16 +285,20 @@ def get_roi_rel_points_test(self, mask_pred, pred_label, cfg):
         num_points = cfg.subdivision_num_points
         uncertainty_map = self._get_uncertainty(mask_pred, pred_label)
         num_rois, _, mask_height, mask_width = uncertainty_map.shape
-        h_step = 1.0 / mask_height
-        w_step = 1.0 / mask_width
-
-        uncertainty_map = uncertainty_map.view(num_rois,
-                                               mask_height * mask_width)
-        num_points = min(mask_height * mask_width, num_points)
+        if isinstance(mask_height, torch.Tensor):
```
[the remark this answers is not in the document]
Comment (AronLin) [reply mmdet/models/roi_heads/mask_heads/mask_point_head.py:288]: I guess that the types of elements in `uncertainty_map.shape` are different in PyTorch and ONNX.
The developer of PR #4977 modified it for exporting ONNX.

File: mmdet/models/roi_heads/mask_heads/mask_point_head.py (hunk 1 above)
Comment (AronLin) [reply mmdet/models/roi_heads/mask_heads/mask_point_head.py:299]: Waston, you found the blind spot!

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 7]
```diff
@@ -9,6 +12,8 @@
 from ..builder import HEADS
 from .standard_roi_head import StandardRoIHead
 
+logger = logging.getLogger(__name__)
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_rend_roi_head.py:15]: Should use get_root_logger() in mmdet.utils

**Comment by AronLin** [comment]
The tests for single image inference, batch inference and ONNX inference  have passed. The PR can be merged now.

File: mmdet/models/roi_heads/point_rend_roi_head.py (hunk 3 above)
Comment (AronLin) [reply mmdet/models/roi_heads/point_rend_roi_head.py:355]: Passes.

> test the batch inference

Passed.

File: mmdet/models/roi_heads/mask_heads/mask_point_head.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jshilong) [reply mmdet/models/roi_heads/mask_heads/mask_point_head.py:288]: Please add comments for this judgement

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 8]
```diff
@@ -207,3 +247,150 @@ def aug_test_mask(self, feats, img_metas, det_bboxes, det_labels):
...
+
+        Args:
+            x (tuple[Tensor]): Feature maps of all scale level.
+            img_metas (list[dict]): Image meta info.
+            det_bboxes (Tensor): Bboxes and corresponding scores.
+                has shape [N, num_bboxes, 5].
+            det_labels (Tensor): class labels of
+                shape [N, num_bboxes].
+
+        Returns:
```
Comment (jshilong) [inline mmdet/models/roi_heads/point_rend_roi_head.py:365]: Please check the return type

File: mmdet/models/roi_heads/point_rend_roi_head.py (hunk 7 above)
Comment (AronLin) [reply mmdet/models/roi_heads/point_rend_roi_head.py:15]: This problem also appears in standard_roi_head, I will solve this problem in another PR.

**Review by jshilong — APPROVED** [review]

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 9]
```diff
@@ -86,7 +104,20 @@ def _get_fine_grained_point_feats(self, x, rois, rel_roi_points,
 
     def _mask_point_forward_test(self, x, rois, label_pred, mask_pred,
                                  img_metas):
-        """Mask refining process with point head in testing."""
+        """Mask refining process with point head in testing.
+
+        Args:
+            x (tuple[Tensor]): Feature maps of all scale level.
+            img_metas (list[dict]): Image meta info.
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_rend_roi_head.py:111]: The order of arguments is different from the interface.

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 10]
```diff
@@ -131,23 +162,32 @@ def simple_test_mask(self,
...
         ori_shapes = tuple(meta['ori_shape'] for meta in img_metas)
         scale_factors = tuple(meta['scale_factor'] for meta in img_metas)
+
+        if isinstance(scale_factors[0], float):
+            logger.warning(
+                'Scale factor in img_metas should be a '
+                'ndarray with shape (4,) '
+                'arrange as (factor_w, factor_h, factor_w, factor_h), '
+                'The scale_factor with float type has been deprecated. ')
+            scale_factors = np.array([scale_factors] * 4, dtype=np.float32)
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_rend_roi_head.py:172]: Why don't we directly make it a torch tensor here?

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 11]
```diff
@@ -207,3 +247,150 @@ def aug_test_mask(self, feats, img_metas, det_bboxes, det_labels):
                 scale_factor=1.0,
                 rescale=False)
         return segm_result
+
+    def _onnx_get_fine_grained_point_feats(self, x, rois, rel_roi_points):
+        """Export the process of sampling fine grained feats to onnx.
+
+        Args:
+            x (tuple[Tensor]): Feature maps of all scale level.
+            img_metas (list[dict]): Image meta info.
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_rend_roi_head.py:256]: arguments order.

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 12]
```diff
@@ -207,3 +247,150 @@ def aug_test_mask(self, feats, img_metas, det_bboxes, det_labels):
...
+                num_rois, channels, num_points)
+            fine_grained_feats.append(point_feats)
+        return torch.cat(fine_grained_feats, dim=1)
+
+    def _mask_point_onnx_export(self, x, rois, label_pred, mask_pred):
+        """Export mask refining process with point head to onnx.
+
+        Args:
+            x (tuple[Tensor]): Feature maps of all scale level.
+            img_metas (list[dict]): Image meta info.
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_rend_roi_head.py:291]: Update docstring.

File: mmdet/models/roi_heads/point_rend_roi_head.py (hunk 10 above)
Comment (AronLin) [reply mmdet/models/roi_heads/point_rend_roi_head.py:172]: 1. In line 224, there is `torch.from_numpy(scale_factor)` which needs `scale_factors` to be np.ndarray.
2. This is copied from `standard_roi_head`.

**Review by ZwwWayne — APPROVED** [review]

**Comment by fschvart** [comment]
Hi, Thanks for all the hard work! 
I've been going insane since I can't make the 3x schedule of PointRend convert to ONNX. Either I'm having issues with the grid_sample, or get other weird errors if I use TensorRT. Is it supposed to work? Am I missing anything?
